mlflow2seldon/api.py

```diff
@@ -0,0 +1,288 @@
+                        else:
+                            registered_model.need_redeploy = False
+                        mlflow_models[registered_model.name] = registered_model
+
+                # deploy models in Seldon
+                for model_name in mlflow_models.keys():
+                    if mlflow_models[model_name].need_redeploy:
+                        await _deploy_model(
+                            mlflow_models[model_name],
+                            neuro_client,
+                            registry_secret_name,
+                        )
+                        mlflow_models[model_name].need_redeploy = False
+
+                # removing outdated models
+                for model_name in set(seldon_models.keys()) - set(mlflow_models.keys()):
+                    _delete_seldon_deployment(seldon_models[model_name])
+
+                # the state is sync
+                seldon_models = mlflow_models.copy()
+                deployed_models = [
+                    f"{n}:{m.model_version}" for n, m in seldon_models.items()
+                ]
+                logging.info(f"Deployed models: {';'.join(deployed_models)}")
+                await asyncio.sleep(DELAY)  # not to overload an API
+
+            except KeyboardInterrupt:
+                logging.warning("Got keyboard interrupt, gracefully shutting down...")
+                break
+            except MlflowException as e:
+                if "Page Not Found" in e.message:
+                    logging.warning(
+                        f"Unable to connect to the MLFlow server. Is it running as a platform job?"
+                    )
+                elif "Sign In" in e.message:
+                    logging.warning(
+                        f"Unable to connect to the MLFlow server - auth is required. SSO is disabled?"
+                    )
+                else:
+                    logging.warning(f"MLFlow server error occured:\n{e}")
+            except Exception as e:
+                logging.warning(f"Unexpected exception (ignoring): {e}")
+    finally:
+        for model in seldon_models.values():
+            _delete_seldon_deployment(model)
+
+
+async def _deploy_model(
+    model: _DeployedModel, neuro_client: Client, registry_secret_name: str
+) -> None:
+    logging.info(f"Deploying model: {model}")
+
+    neuro_token = await neuro_client.config.token()
+    deployment_json = _create_seldon_deployment(
+        name=model.name,
+        namespace=model.deployment_namespace,
+        neuro_login_token=neuro_token,
+        neuro_cluster=neuro_client.config.cluster_name,
+        registry_secret_name=registry_secret_name,
+        model_image_ref=model.image,
+        model_storage_uri=str(model.model_storage_uri),
+    )
+    deployment = yaml.dump(deployment_json)
+    path = Path(tempfile.mktemp())
+    path.write_text(deployment)
+    try:
+        subprocess.run(f"kubectl apply -f {path}", shell=True, check=True)
```

Comment:
if this service fails and gets restarted, will we re-deploy the model? as a potential improvement we could store the current model's version not in-memory but persist it -- for example in a ConfigMap, WDYT?

Reply:
Didn't test it, I suppose, the k8s will return an error, will check and report. Good catch!

Reply:
well theoretically K8s will do nothing -- you apply existing the same object... but yeah.  Not critical for demos.